src/building/optimized-build.md

```diff
@@ -0,0 +1,131 @@
+# Optimized build of the compiler
+
+<!-- toc -->
+
+There are multiple additional build configuration options and techniques that can used to compile a
+build of `rustc` that is as optimized as possible (for example when building `rustc` for a Linux
+distribution). The status of these configuration options for various Rust targets is tracked [here].
+This page describes how you can use these approaches when building `rustc` yourself.
+
+[here]: https://github.com/rust-lang/rust/issues/103595
+
+## Link-time optimization
+
+Link-time optimization is a powerful compiler technique that can increase program performance. To
+enable (Thin-)LTO when building `rustc`, set the `rust.lto` config option to `"thin"`
+in `config.toml`:
+
+```toml
+[rust]
+lto = "thin"
+```
+
+> Note that LTO for `rustc` is currently supported and tested only for
+> the `x86_64-unknown-linux-gnu` target. Other targets *may* work, but no guarantees are provided.
+> Notably, LTO optimized `rustc` currently produces [miscompilations] on Windows.
+
+[miscompilations]: https://github.com/rust-lang/rust/issues/109114
+
+Enabling LTO on Linux has [produced] speed-ups by up to 10%.
+
+[produced]: https://github.com/rust-lang/rust/pull/101403#issuecomment-1288190019
+
+## Memory allocator
+
+Using a different memory allocator for `rustc` can provide significant performance benefits. If you
+want to enable the `jemalloc` allocator, you can set the `rust.jemalloc` option to `true`
+in `config.toml`:
+
+```toml
+[rust]
+jemalloc = true
+```
+
+> Note that this option is currently only supported for Linux and macOS targets.
+
+## Codegen units
+
+Reducing the amount of codegen units per `rustc` crate can produce a faster build of the compiler.
+You can modify the number of codegen units for `rustc` and `libstd` in `config.toml` with the
+following options:
+
+```toml
+[rust]
+codegen-units = 1
+codegen-units-std = 1
+```
+
+## Instruction set
+
+By default, `rustc` is compiled for a generic (and conservative) instruction set architecture
+(depending on the selected target), to make it support as many CPUs as possible. If you want to
+compile `rustc` for a specific instruction set architecture, you can set the `target_cpu` compiler
+option in `RUSTFLAGS`:
+
+```bash
+$ RUSTFLAGS="-C target_cpu=x86-64-v3" x.py build ...
+```
+
+If you also want to compile LLVM for a specific instruction set, you can set `llvm` flags
+in `config.toml`:
+
+```toml
+[llvm]
+cxxflags = "-march=x86-64-v3"
+cflags = "-march=x86-64-v3"
+```
+
+## Profile-guided optimization
+
+Applying profile-guided optimizations (or more generally, feedback-directed optimizations) can
+produce a large increase to `rustc` performance, by up to 25%. However, these techniques are not
```

Comment:
can we have a link to some benchmark

Reply:
PGO on Linux was enabled in parts (rustc/LLVM), so I don't really have a single benchmark link that would such big performance wins at once. This number comes from my past experience, plus from some attempts to disable PGO (so I saw the opposite regressions). I have changed the amount to `up to 15%` and added a link to PGO for Windows: https://github.com/rust-lang/rust/pull/96978, which enabled PGO for `rustc` and LLVM at once, and also to https://blog.rust-lang.org/inside-rust/2020/11/11/exploring-pgo-for-the-rust-compiler.html#final-numbers-and-a-benchmarking-plot-twist.